src/game_battler.cpp

```diff
@@ -38,9 +38,8 @@ Game_Battler::Game_Battler() {
 }
 
 bool Game_Battler::HasState(int state_id) const {
-	const std::vector<int16_t> states = GetInflictedStates();
-
-	return (std::find(states.begin(), states.end(), state_id) != states.end());
+	auto& states = GetStates();
+	return state_id >= 1 && state_id <= states.size() && states[state_id-1] > 0;
```

Comment:
We also have a sanity checking function array access function in ReaderUtil:
```
ReaderUtil::GetElement(states, state_id);
```
Is that suitable here?

Reply:
If the `states` vector was not the same size as the `Data::states` then using `GetElement()` would be wrong. It could even result in a segfault.

My latest liblcf patch resizes `states` to always match `Data::states` so probably `GetElement()` is ok now, but maybe we will change the design later?

Doing it this way at least is bulletproof.

Reply:
That's why I passed "states" (which is the local variable in yor code) to GetElement, not "Data::states".
Also you don't resize the states array of enemies, so an unchecked state access will still segfault (we had this before).

Reply:
Oh I see now. I didn't realize just how general `GetElement()` is.

Yes I agree with you then, this is much better.